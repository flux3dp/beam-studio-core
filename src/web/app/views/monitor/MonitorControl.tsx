import React, { useContext } from 'react';
import { Button, Space } from 'antd';
import { PauseCircleFilled, PlayCircleFilled, StopFilled } from '@ant-design/icons';

import DeviceConstants from 'app/constants/device-constants';
import i18n from 'helpers/i18n';
import MonitorStatus, { ButtonTypes } from 'helpers/monitor-status';
import { useIsMobile } from 'helpers/system-helper';
import { Mode } from 'app/constants/monitor-constants';
import { MonitorContext } from 'app/contexts/MonitorContext';

const LANG = i18n.lang.monitor;

const MonitorControl = (): JSX.Element => {
  const isMobile = useIsMobile();
  const buttonShape = isMobile ? 'round' : 'default';
  const { onPlay, onPause, onStop, mode, report } = useContext(MonitorContext);
  const mapButtonTypeToElement = (type: ButtonTypes): JSX.Element => {
    const enabled = type % 2 === 1;
    switch (type) {
      case ButtonTypes.PLAY:
      case ButtonTypes.DISABLED_PLAY:
        return (
          <Button
            key={type}
            disabled={!enabled}
            shape={buttonShape}
            type="primary"
            onClick={onPlay}
          >
            <PlayCircleFilled />
            {(report.st_id === DeviceConstants.status.PAUSED) ? LANG.resume : LANG.go}
          </Button>
        );
      case ButtonTypes.PAUSE:
      case ButtonTypes.DISABLED_PAUSE:
        return (
          <Button
            key={type}
            disabled={!enabled}
            shape={buttonShape}
            type="primary"
            onClick={onPause}
          >
            <PauseCircleFilled />
            {LANG.pause}
          </Button>
        );
      case ButtonTypes.STOP:
      case ButtonTypes.DISABLED_STOP:
        return (
          <Button key={type} disabled={!enabled} shape={buttonShape} onClick={onStop}>
            <StopFilled />
            {LANG.stop}
          </Button>
        );
      default:
        return null;
    }
  };

  // const renderRelocateButton = (): JSX.Element => {
  //   const { isMaintainMoving, onRelocate } = this.context;
  //   const className = classNames('controls right', { disabled: isMaintainMoving });
  //   return (
  //     <div className={className} onClick={onRelocate}>
  //       <div className="btn-control btn-relocate">
  //         <img src="img/beambox/icon-target.svg" />
  //       </div>
  //       <div className="description">{LANG.relocate}</div>
  //     </div>
  //   );
  // // }

  const canStart = !!report && report.st_id === DeviceConstants.status.IDLE;

  if (mode === Mode.PREVIEW || mode === Mode.FILE_PREVIEW) {
    return (
      <Space>
        <Button disabled={!canStart} shape={buttonShape} type="primary" onClick={onPlay}>
          <PlayCircleFilled />
          {LANG.go}
        </Button>
      </Space>
    );
  }
  if (mode === Mode.WORKING) {
    return (
      <Space>
        {MonitorStatus.getControlButtonType(report).map((v) => mapButtonTypeToElement(v))}
      </Space>
    );
  }
  return <div />;
};

export default MonitorControl;
