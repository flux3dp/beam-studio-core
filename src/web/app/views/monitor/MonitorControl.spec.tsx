/* eslint-disable import/first */
import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('helpers/i18n', () => ({
  lang: {
    monitor: {
      go: 'Start',
      pause: 'Pause',
      stop: 'Stop',
      camera: 'Camera',
      upload: 'Upload',
      download: 'Download',
      relocate: 'Relocate',
      cancel: 'Cancel',
    },
  },
}));

jest.mock('app/contexts/Monitor-Context', () => ({
  MonitorContext: React.createContext(null),
}));

import MonitorControl from './MonitorControl';

describe('should render correctly', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('mode is FILE', () => {
    const onUpload = jest.fn();
    const onDownload = jest.fn();
    const toggleCamera = jest.fn();
    MonitorControl.contextType = React.createContext({
      mode: 'FILE',
      isMaintainMoving: false,
      currentPath: [],
      highlightedItem: {},
      report: {
        st_id: 1,
      },
      onUpload,
      toggleCamera,
      onRelocate: jest.fn(),
      onPlay: jest.fn(),
      onPause: jest.fn(),
      onStop: jest.fn(),
      onDownload,
      endRelocate: jest.fn(),
    });
    const wrapper = mount(<MonitorControl />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('div.controls.left').simulate('click');
    expect(onUpload).toHaveBeenCalledTimes(1);
    wrapper.find('input.upload-control').simulate('change');
    expect(onUpload).toHaveBeenCalledTimes(2);

    wrapper.find('div.controls.center').simulate('click');
    expect(onDownload).toHaveBeenCalledTimes(1);

    wrapper.find('div.controls.right').simulate('click');
    expect(toggleCamera).toHaveBeenCalledTimes(1);
  });

  test('mode is PREVIEW', () => {
    MonitorControl.contextType = React.createContext({
      mode: 'PREVIEW',
      isMaintainMoving: false,
      currentPath: [],
      highlightedItem: {},
      report: {
        st_id: 0,
      },
      onUpload: jest.fn(),
      toggleCamera: jest.fn(),
      onRelocate: jest.fn(),
      onPlay: jest.fn(),
      onPause: jest.fn(),
      onStop: jest.fn(),
      onDownload: jest.fn(),
      endRelocate: jest.fn(),
    });
    const wrapper = mount(<MonitorControl />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('mode is FILE_PREVIEW', () => {
    MonitorControl.contextType = React.createContext({
      mode: 'FILE_PREVIEW',
      isMaintainMoving: false,
      currentPath: [],
      highlightedItem: {},
      report: {
        st_id: 1,
      },
      onUpload: jest.fn(),
      toggleCamera: jest.fn(),
      onRelocate: jest.fn(),
      onPlay: jest.fn(),
      onPause: jest.fn(),
      onStop: jest.fn(),
      onDownload: jest.fn(),
      endRelocate: jest.fn(),
    });
    const wrapper = mount(<MonitorControl />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('mode is WORKING', () => {
    test('ButtonTypes.DISABLED_STOP and ButtonTypes.DISABLED_PLAY', () => {
      MonitorControl.contextType = React.createContext({
        mode: 'WORKING',
        isMaintainMoving: false,
        currentPath: [],
        highlightedItem: {},
        report: {},
        onUpload: jest.fn(),
        toggleCamera: jest.fn(),
        onRelocate: jest.fn(),
        onPlay: jest.fn(),
        onPause: jest.fn(),
        onStop: jest.fn(),
        onDownload: jest.fn(),
        endRelocate: jest.fn(),
      });
      const wrapper = mount(<MonitorControl />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('ButtonTypes.STOP and ButtonTypes.PAUSE', () => {
      MonitorControl.contextType = React.createContext({
        mode: 'WORKING',
        isMaintainMoving: false,
        currentPath: [],
        highlightedItem: {},
        report: {
          st_id: 16,
        },
        onUpload: jest.fn(),
        toggleCamera: jest.fn(),
        onRelocate: jest.fn(),
        onPlay: jest.fn(),
        onPause: jest.fn(),
        onStop: jest.fn(),
        onDownload: jest.fn(),
        endRelocate: jest.fn(),
      });
      const wrapper = mount(<MonitorControl />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('ButtonTypes.STOP and ButtonTypes.PLAY', () => {
      MonitorControl.contextType = React.createContext({
        mode: 'WORKING',
        isMaintainMoving: false,
        currentPath: [],
        highlightedItem: {},
        report: {
          st_id: 32,
        },
        onUpload: jest.fn(),
        toggleCamera: jest.fn(),
        onRelocate: jest.fn(),
        onPlay: jest.fn(),
        onPause: jest.fn(),
        onStop: jest.fn(),
        onDownload: jest.fn(),
        endRelocate: jest.fn(),
      });
      const wrapper = mount(<MonitorControl />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('ButtonTypes.DISABLED_STOP and ButtonTypes.DISABLED_PAUSE', () => {
      MonitorControl.contextType = React.createContext({
        mode: 'WORKING',
        isMaintainMoving: false,
        currentPath: [],
        highlightedItem: {},
        report: {
          st_id: 130,
        },
        onUpload: jest.fn(),
        toggleCamera: jest.fn(),
        onRelocate: jest.fn(),
        onPlay: jest.fn(),
        onPause: jest.fn(),
        onStop: jest.fn(),
        onDownload: jest.fn(),
        endRelocate: jest.fn(),
      });
      const wrapper = mount(<MonitorControl />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  test('mode is CAMERA', () => {
    MonitorControl.contextType = React.createContext({
      mode: 'CAMERA',
      isMaintainMoving: false,
      currentPath: [],
      highlightedItem: {},
      report: {},
      onUpload: jest.fn(),
      toggleCamera: jest.fn(),
      onRelocate: jest.fn(),
      onPlay: jest.fn(),
      onPause: jest.fn(),
      onStop: jest.fn(),
      onDownload: jest.fn(),
      endRelocate: jest.fn(),
    });
    const wrapper = mount(<MonitorControl />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('mode is CAMERA_RELOCATE', () => {
    const onRelocate = jest.fn();
    MonitorControl.contextType = React.createContext({
      mode: 'CAMERA_RELOCATE',
      isMaintainMoving: false,
      currentPath: [],
      highlightedItem: {},
      report: {},
      onUpload: jest.fn(),
      toggleCamera: jest.fn(),
      onRelocate,
      onPlay: jest.fn(),
      onPause: jest.fn(),
      onStop: jest.fn(),
      onDownload: jest.fn(),
      endRelocate: jest.fn(),
    });
    const wrapper = mount(<MonitorControl />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('div.controls.right').simulate('click');
    expect(onRelocate).toHaveBeenCalledTimes(1);
  });
});
