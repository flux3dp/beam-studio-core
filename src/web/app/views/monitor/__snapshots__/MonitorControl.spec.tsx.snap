// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly mode is CAMERA 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-go"
      />
      <div
        className="description"
      >
        Start
      </div>
    </div>
    <div
      className="controls right disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control on"
      />
      <div
        className="description on"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is CAMERA_RELOCATE 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-cancel"
      />
      <div
        className="description"
      >
        Cancel
      </div>
    </div>
    <div
      className="controls right"
      onClick={[MockFunction]}
    >
      <div
        className="btn-control btn-relocate"
      >
        <img
          src="img/beambox/icon-target.svg"
        />
      </div>
      <div
        className="description"
      >
        Relocate
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is FILE 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-upload btn-control"
      />
      <input
        accept=".fc"
        className="upload-control"
        onChange={[MockFunction]}
        type="file"
      />
      <div
        className="description"
      >
        Upload
      </div>
    </div>
    <div
      className="controls center disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-download"
      />
      <div
        className="description"
      >
        Download
      </div>
    </div>
    <div
      className="controls right disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is FILE_PREVIEW 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-go"
      />
      <div
        className="description"
      >
        Start
      </div>
    </div>
    <div
      className="controls right disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is PREVIEW 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-go"
      />
      <div
        className="description"
      >
        Start
      </div>
    </div>
    <div
      className="controls right"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is WORKING ButtonTypes.DISABLED_STOP and ButtonTypes.DISABLED_PAUSE 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-pause"
      />
      <div
        className="description"
      >
        Pause
      </div>
    </div>
    <div
      className="controls right disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is WORKING ButtonTypes.DISABLED_STOP and ButtonTypes.DISABLED_PLAY 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center disabled"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-go"
      />
      <div
        className="description"
      >
        Start
      </div>
    </div>
    <div
      className="controls right disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is WORKING ButtonTypes.STOP and ButtonTypes.PAUSE 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-pause"
      />
      <div
        className="description"
      >
        Pause
      </div>
    </div>
    <div
      className="controls right disabled"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;

exports[`should render correctly mode is WORKING ButtonTypes.STOP and ButtonTypes.PLAY 1`] = `
<MonitorControl>
  <div
    className="operation"
  >
    <div
      className="controls left"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-stop"
      />
      <div
        className="description"
      >
        Stop
      </div>
    </div>
    <div
      className="controls center"
      onClick={[Function]}
    >
      <div
        className="btn-control btn-go"
      />
      <div
        className="description"
      >
        Start
      </div>
    </div>
    <div
      className="controls right"
      onClick={[MockFunction]}
    >
      <div
        className="btn-camera btn-control"
      />
      <div
        className="description"
      >
        Camera
      </div>
    </div>
  </div>
</MonitorControl>
`;
